design/authenticators-status/authn_status_general.md

```diff
@@ -0,0 +1,113 @@
+and those specific to each authenticator:
+
+- The requesting user has access to the authenticator status route webservice
+    - more info in the Security section of the [epic](https://github.com/cyberark/conjur/issues/1062)
+- The authenticator is implemented
+- The account exists
+    - although a wrong account in an `/authenticate` request will indicate a client error and
+    not a server error, we still need to verify it exists in order to verify that the webservice exists
+    under that account
+- The webservice exists
+- The authenticator is enabled in the ENV
+- [Specific authenticator requirements](#implementing_specific_authenticator_status_check)
+ 
+***Note:*** The default authenticator (`authn`) is always healthy
+
+We will also need to audit access success & failures to this endpoint.
+ 
+## Implementing specific authenticator status check
+ 
+Some authenticators need extra validation. As mentioned in the Epic, if 
+the status check is not implemented for the authenticator we will return a 503 Not 
+Implemented response. 
+
+In the following section, we will use `authn-oidc` as an example.
+ 
+For each authenticator that will implement the status check, we will add:
+ 1. A `Status` CommandClass in the structure `Authentication::AuthnOidc::Status` (with the validations in its `call` method). 
+ 1. A `status` method to its existing authenticator class (i.e. `Authentication::AuthnOidc::Authenticator`)
+ which will call the new CommandClass, as follows: 
+ 
+ ```
+ def status
+   Authentication::AuthnOidc::Status.new.(
+     <input for status check>
+   )
+ end
+ ``` 
+ 
+ The `status` method above will be called from the general status check, when validating
+ the authenticator's specific requirements. We will first verify that the `status` method exists
+ in the `Authenticator` class, which will indicate that the status check is implemented on the given
+ authenticator. Therefore, the `call` method of the general status check 
+ will look like the following:
+                                                                            
+ ```
+ def call
+  validate_authenticator_exists
+  validate_authenticator_implements_status_check
+  .
+  .
+  # perform general validations (whitelisted in env, etc.)
+  .
+  .
+  validate_authenticator_requirements
+ end
+ 
+ private
+ 
+ def validate_authenticator_exists
+   raise Err::AuthenticatorNotFound unless authenticator
+ end
+     
+ def validate_authenticator_implements_status_check
+  unless authenticator.method_defined?(:status)
+    raise 503
+ end
+ 
+ def validate_authenticator_requirements
+  authenticator.status
+ end
+ 
+ def authenticator
+  # The `@authenticators` map includes all authenticator classes that are implemented in 
+  # Conjur (`Authentication::AuthnOidc::Authenticator`, `Authentication::AuthnLdap::Authenticator`, etc.). 
+  
+  @authenticator = @authenticators[@authenticator_input.authenticator_name]
+ end
+ ```
+ 
+ # Test Plan
```

Comment:
@orenbm  I believe we need to add a test for a valid authenticator and a valid access token but the authenticator returns status false then the result should be 500?
what would be the text of other fields:  "error": "<configuration_error>",
    "message": "<conjur_error_message>",

Reply:
@orenbm i believe that the Then part of all the other test case should also include  "error": "<configuration_error>",
    "message": "<conjur_error_message>",

Reply:
absolutely! will add this to the doc

Reply:
> @orenbm I believe we need to add a test for a valid authenticator and a valid access token but the authenticator returns status false then the result should be 500?
> what would be the text of other fields: "error": "<configuration_error>",
> "message": "<conjur_error_message>",

The message will be the message of the relevant error. if the status check failed on `WebserviceNotFound` then the message will be `Webservice '{0-webservice-name}' wasn't found` (which is its built-in message)